File: execution/backtest_adapter.py
from __future__ import annotations

import uuid
from datetime import datetime

from core.signal import Direction, Signal
from execution.base import ExecutionAdapter, Trade


class BacktestAdapter(ExecutionAdapter):
    def __init__(self, initial_balance: float = 10000.0, risk_per_trade: float = 0.01) -> None:
        self.balance = initial_balance
        self.risk_per_trade = risk_per_trade
        self._open_positions: dict[str, Trade] = {}
        self._closed_trades: list[Trade] = []

    def place_order(self, signal: Signal) -> str:
        order_id = str(uuid.uuid4())
        size = (self.balance * self.risk_per_trade) / signal.risk_amount
        trade = Trade(
            signal=signal,
            order_id=order_id,
            entry_time=signal.timestamp,
            entry_price=signal.entry,
            position_size=size,
        )
        self._open_positions[signal.symbol] = trade
        return order_id

    def get_position(self, symbol: str) -> dict | None:
        trade = self._open_positions.get(symbol)
        if trade is None:
            return None
        return {
            "symbol": symbol,
            "order_id": trade.order_id,
            "entry_price": trade.entry_price,
            "position_size": trade.position_size,
            "direction": trade.signal.direction,
        }

    def close_position(self, symbol: str) -> None:
        trade = self._open_positions.get(symbol)
        if trade is None:
            return
        self._close_trade(symbol, trade.signal.stop_loss, datetime.utcnow(), "manual")

    def update(
        self,
        symbol: str,
        candle_high: float,
        candle_low: float,
        candle_time: datetime,
        candle_close: float | None = None,
    ) -> None:
        trade = self._open_positions.get(symbol)
        if trade is None:
            return

        sl = trade.signal.stop_loss
        tp = trade.signal.take_profit
        direction = trade.signal.direction

        if direction == Direction.LONG:
            hit_tp = candle_high >= tp
            hit_sl = candle_low <= sl
        else:
            hit_tp = candle_low <= tp
            hit_sl = candle_high >= sl

        if hit_tp:
            self._close_trade(symbol, tp, candle_time, "tp")
            return
        if hit_sl:
            self._close_trade(symbol, sl, candle_time, "sl")
            return

        expiry = trade.signal.expiry_time
        if expiry is not None and candle_time >= expiry:
            price = candle_close if candle_close is not None else (candle_high + candle_low) / 2
            self._close_trade(symbol, price, candle_time, "expiry")

    @property
    def trades(self) -> list[Trade]:
        return list(self._closed_trades)

    @property
    def open_trades(self) -> list[Trade]:
        return list(self._open_positions.values())

    def report(self) -> dict:
        closed = self._closed_trades
        if not closed:
            return {
                "total_trades": 0,
                "winners": 0,
                "losers": 0,
                "win_rate": 0.0,
                "profit_factor": 0.0,
                "total_pnl_pct": 0.0,
                "max_drawdown_pct": 0.0,
            }

        winners = [t for t in closed if t.is_winner]
        losers = [t for t in closed if not t.is_winner]

        gross_profit = sum(t.pnl_pct for t in winners if t.pnl_pct is not None)
        gross_loss = abs(sum(t.pnl_pct for t in losers if t.pnl_pct is not None))
        profit_factor = gross_profit / gross_loss if gross_loss > 0 else float("inf")

        cumulative = 0.0
        peak = 0.0
        max_dd = 0.0
        for t in closed:
            pnl = t.pnl_pct or 0.0
            cumulative += pnl
            if cumulative > peak:
                peak = cumulative
            dd = peak - cumulative
            if dd > max_dd:
                max_dd = dd

        return {
            "total_trades": len(closed),
            "winners": len(winners),
            "losers": len(losers),
            "win_rate": len(winners) / len(closed),
            "profit_factor": profit_factor,
            "total_pnl_pct": sum(t.pnl_pct for t in closed if t.pnl_pct is not None),
            "max_drawdown_pct": max_dd,
        }

    def _close_trade(self, symbol: str, price: float, time: datetime, reason: str) -> None:
        trade = self._open_positions.pop(symbol, None)
        if trade is None:
            return
        trade.exit_price = price
        trade.exit_time = time
        trade.exit_reason = reason
        if trade.signal.direction == Direction.LONG:
            dollar_pnl = trade.position_size * (price - trade.entry_price)
        else:
            dollar_pnl = trade.position_size * (trade.entry_price - price)
        self.balance += dollar_pnl
        self._closed_trades.append(trade)

File: src/execution/backtest.adapter.ts
import { ExecutionPort, Position } from '../domain/ports';
import { Direction, Signal } from '../domain/signal';
import { ExitReason, Trade } from '../domain/trade';

export interface BacktestReport {
  totalTrades: number;
  winners: number;
  losers: number;
  winRate: number;
  profitFactor: number;
  totalPnlPct: number;
  maxDrawdownPct: number;
}

export interface BacktestAdapterOptions {
  readonly initialBalance?: number;
  readonly riskPerTrade?: number;
}

/**
 * Fill simulation for backtests.
 *
 * `placeOrder` opens a trade immediately at `signal.entry`; each `update()`
 * checks one bar's high/low against the signal's levels. The order of those
 * checks — take profit, then stop loss, then expiry — is deliberate and
 * load-bearing: when a single bar spans both TP and SL there is no way to know
 * which came first intraday, and the Python implementation resolved that in
 * favour of the target. Reordering the checks would silently change the
 * outcome of every such bar in the historical record.
 *
 * Note this sizes positions itself rather than going through `RiskManager`,
 * which is what the Python `BacktestAdapter` did.
 */
export class BacktestAdapter implements ExecutionPort {
  balance: number;
  readonly riskPerTrade: number;

  private readonly openPositions = new Map<string, Trade>();
  private readonly closedTrades: Trade[] = [];

  constructor(options: BacktestAdapterOptions = {}) {
    this.balance = options.initialBalance ?? 10_000;
    this.riskPerTrade = options.riskPerTrade ?? 0.01;
  }

  placeOrder(signal: Signal): string {
    // The Web Crypto global rather than `node:crypto`, so this layer stays
    // free of platform imports.
    const orderId = globalThis.crypto.randomUUID();
    const size = (this.balance * this.riskPerTrade) / signal.riskAmount;

    this.openPositions.set(
      signal.symbol,
      new Trade({
        signal,
        orderId,
        entryTime: signal.timestamp,
        entryPrice: signal.entry,
        positionSize: size,
      }),
    );
    return orderId;
  }

  getPosition(symbol: string): Position | null {
    const trade = this.openPositions.get(symbol);
    if (trade === undefined) {
      return null;
    }
    return {
      symbol,
      orderId: trade.orderId,
      entryPrice: trade.entryPrice,
      positionSize: trade.positionSize,
      direction: trade.signal.direction,
    };
  }

  /**
   * Discretionary close at the stop level.
   *
   * The backtest loop never calls this — it is here so the port is fully
   * implemented. It stamps the wall clock, the one non-deterministic thing in
   * this class, exactly as the Python version's `datetime.utcnow()` did.
   */
  closePosition(symbol: string): void {
    const trade = this.openPositions.get(symbol);
    if (trade === undefined) {
      return;
    }
    this.closeTrade(symbol, trade.signal.stopLoss, Date.now(), 'manual');
  }

  /** Settle the open position for `symbol` against one bar. */
  update(
    symbol: string,
    candleHigh: number,
    candleLow: number,
    candleTime: number,
    candleClose: number | null = null,
  ): void {
    const trade = this.openPositions.get(symbol);
    if (trade === undefined) {
      return;
    }

    const { stopLoss, takeProfit, direction, expiryTime } = trade.signal;

    const hitTakeProfit =
      direction === Direction.Long ? candleHigh >= takeProfit : candleLow <= takeProfit;
    const hitStopLoss =
      direction === Direction.Long ? candleLow <= stopLoss : candleHigh >= stopLoss;

    if (hitTakeProfit) {
      this.closeTrade(symbol, takeProfit, candleTime, 'tp');
      return;
    }
    if (hitStopLoss) {
      this.closeTrade(symbol, stopLoss, candleTime, 'sl');
      return;
    }

    if (expiryTime !== null && candleTime >= expiryTime) {
      const price = candleClose !== null ? candleClose : (candleHigh + candleLow) / 2;
      this.closeTrade(symbol, price, candleTime, 'expiry');
    }
  }

  get trades(): Trade[] {
    return [...this.closedTrades];
  }

  get openTrades(): Trade[] {
    return [...this.openPositions.values()];
  }

  report(): BacktestReport {
    const closed = this.closedTrades;
    if (closed.length === 0) {
      return {
        totalTrades: 0,
        winners: 0,
        losers: 0,
        winRate: 0,
        profitFactor: 0,
        totalPnlPct: 0,
        maxDrawdownPct: 0,
      };
    }

    const winners = closed.filter((trade) => trade.isWinner);
    const losers = closed.filter((trade) => !trade.isWinner);

    // Summation order matches the Python original bar for bar. Python 3.11's
    // sum() accumulates naively, like this reduce; 3.12 switched to
    // compensated summation, which is why the golden fixtures pin 3.11.
    let grossProfit = 0;
    for (const trade of winners) {
      const pnl = trade.pnlPct;
      if (pnl !== null) {
        grossProfit += pnl;
      }
    }
    let grossLossSigned = 0;
    for (const trade of losers) {
      const pnl = trade.pnlPct;
      if (pnl !== null) {
        grossLossSigned += pnl;
      }
    }
    const grossLoss = Math.abs(grossLossSigned);

    let cumulative = 0;
    let peak = 0;
    let maxDrawdown = 0;
    let totalPnlPct = 0;
    for (const trade of closed) {
      const pnl = trade.pnlPct ?? 0;
      cumulative += pnl;
      if (cumulative > peak) {
        peak = cumulative;
      }
      const drawdown = peak - cumulative;
      if (drawdown > maxDrawdown) {
        maxDrawdown = drawdown;
      }
    }
    for (const trade of closed) {
      const pnl = trade.pnlPct;
      if (pnl !== null) {
        totalPnlPct += pnl;
      }
    }

    return {
      totalTrades: closed.length,
      winners: winners.length,
      losers: losers.length,
      winRate: winners.length / closed.length,
      profitFactor: grossLoss > 0 ? grossProfit / grossLoss : Number.POSITIVE_INFINITY,
      totalPnlPct,
      maxDrawdownPct: maxDrawdown,
    };
  }

  private closeTrade(
    symbol: string,
    price: number,
    time: number,
    reason: ExitReason,
  ): void {
    const trade = this.openPositions.get(symbol);
    if (trade === undefined) {
      return;
    }
    this.openPositions.delete(symbol);

    trade.exitPrice = price;
    trade.exitTime = time;
    trade.exitReason = reason;

    const dollarPnl =
      trade.signal.direction === Direction.Long
        ? trade.positionSize * (price - trade.entryPrice)
        : trade.positionSize * (trade.entryPrice - price);

    this.balance += dollarPnl;
    this.closedTrades.push(trade);
  }
}
